### ISSUE TITLE
Process getting stuck at 100% with vips 8.11

### ISSUE THREAD

**Issue by adityapatadia** [libvips/libvips#2430]
I would use proper steps to reproduce but I am not able to perfectly reproduce. We use libvips with sharp in production. The process keeps running all the time.

When we deploy 8.11 version to production, at a random time, some of our machines go to 100% cpu usage and then sticks there for very long time of about 5-6 hours. We then have to close those processed (pods in our kubernetes) and then have to redeploy them.

When we just downgraded vips to 8.10.6 and sharp to 0.28.3, the issues went away.

I am ready to provide whatever information you need.

**Comment by jcupitt** [comment]
Hi @adityapatadia, oh dear, that's bad.

Have you tried 8.11.3? It has some fixes to the threading system which might help.

https://github.com/libvips/libvips/releases/tag/v8.11.3

**Comment by adityapatadia** [comment]
Yes 8.11.3 has this issue. I am not sure what can I do to reproduce it on local machine but any steps you recommend to get correct data from production, I will provide.

I believe it has something to do with the threadpool that is introduced in 8.11z

—
Regards,
Aditya Patadia

**Comment by jcupitt** [comment]
That's bad news :( I agree it sounds like perhaps a livelock and it could be an issue with the new threadpool. These problems can be very hard to diagnose.

What's your platform?

@lovell and @kleisauke have you seen anything like this?

**Comment by kleisauke** [comment]
I haven't seen anything like this under heavy load. The only bug I'm aware of regarding the threadpool was fixed with libvips 8.11.3 and only affected macOS.

FWIW, I tried a variant of the test case discussed in comment https://github.com/libvips/php-vips-ext/issues/42#issuecomment-879817757 with a hardened Alpine Docker image.

<details>
  <summary>Details</summary>
 
`test-2430.c`:
```c
// compile with
// gcc -g -Wall test-2430.c `pkg-config vips --cflags --libs` -o test-2430
// run with
// ./test-2430
#include <sys/wait.h>
#include <vips/vips.h>

void work() {
  VipsImage *im;
  double avg;

  if (vips_black(&im, 100, 100, NULL))
    vips_error_exit("vips_black failed");
  if (vips_avg(im, &avg, NULL)) {
    g_object_unref(im);
    vips_error_exit("vips_avg failed");
  }
  g_object_unref(im);

  printf("avg = %f\n", avg);
}

int main(int argc, char **argv) {
  pid_t pid;

  // Disable buffering on stdout
  setbuf(stdout, NULL);

  printf("vips_init\n");
  if (VIPS_INIT(argv[0]))
    vips_error_exit("unable to start VIPS");

  // Initial work, this starts the GLib threadpool
  work();

  // Fork the process, any other threads will disappear
  printf("fork\n");
  pid = fork();

  switch (pid) {
    case -1:
      vips_error_exit("fork failed with: %s", strerror(errno));

    case 0:
      printf("PID %d (child) doing work\n", getpid());

      work();

      printf("child done\n");
      break;

    default:
      printf("PID %d (parent) waiting for child PID %d\n", getpid(), pid);
      waitpid(pid, NULL, 0);
      
      work();

      printf("exit\n");
      exit(0);
  }

  return 0;
}
```

```bash
# Alpine edge provides vips 8.11.3
$ docker run --cap-drop ALL --cap-add CHOWN -v $(pwd):/app -w /app -it alpine:edge sh -c "\
  apk add build-base vips-dev && \
  gcc -g -Wall test-2430.c -I/usr/lib/glib-2.0/include `pkg-config vips --cflags --libs` -o test-2430 && \
  G_MESSAGES_DEBUG=all ./test-2430"
```

Output:
```
vips_init
VIPS-INFO: 13:11:39.339: found /usr/lib/vips-modules-8.11
VIPS-INFO: 13:11:39.339: using configure-time prefix
VIPS-INFO: 13:11:39.339: VIPS_PREFIX = /usr
VIPS-INFO: 13:11:39.339: VIPS_LIBDIR = /usr/lib
VIPS-INFO: 13:11:39.339: prefix = /usr
VIPS-INFO: 13:11:39.339: libdir = /usr/lib
VIPS-INFO: 13:11:39.339: searching "/usr/lib/vips-modules-8.11"
VIPS-INFO: 13:11:39.340: loading "/usr/lib/vips-modules-8.11/vips-heif.so"
VIPS-INFO: 13:11:39.344: searching "/usr/lib/vips-plugins-8.11"
avg = 0.000000
fork
PID 1 (parent) waiting for child PID 37
PID 37 (child) doing work
avg = 0.000000
child done
avg = 0.000000
exit
```
</details>

But couldn't reproduce that anymore, it seems that even a `fork()` after the GLib threadpool is started will work without any problems.

I've a couple of questions:
- Are you using the latest sharp version with the provided pre-built libvips binaries? If so, ignore the next questions.
- Was libvips 8.11.3 build with the JP2K loader and/or saver enabled?
- Was libvips 8.11.3 build with loadable modules support?
- Did you ensure that this is not the same issue as what you reported in https://github.com/lovell/sharp/issues/2802?

**Comment by lovell** [comment]
I've not seen this, nor have had any other reports of this.

Is this using prebuilt binaries? If not, what versions of all dependencies are involved? What CPU/OS? What concurrency settings? What image formats?

If this is on Linux then use of the `perf` command line tool might help reveal what the CPU is doing.

**Comment by jcupitt** [comment]
You're both right, there are loads of other possible causes. Your server process leaking and wobbling off into swap death, for example. Does this tend to happen with certain image types, or during certain classes of operation?

We will need some way of reproducing this before we can investigate, unfortunately.

**Comment by adityapatadia** [comment]
@kleisauke I am using latest sharp version but I build libvips on my own.
* I haven't built it with JP2K
* I am not sure if I have built libvips with loadable modules support. How can I check?
* The SVG issue was the one I myself raised and once it was identified, I upgraded my librsvg to version that claims to fix it.

@lovell I have tried perf command but did not find anything. I will mail you the output.

I am building libvips with MOZJPEG-4.0.2, Libwebp-1.2.1, Libaom-3.1.2, Libheif-1.12.0 and libspng-0.6.3. I will try to bring more data.

We are processing more than 100 million images everyday and that is making it hard for us to reproduce the issue with single image. We are not sure what is causing it. However, I am certain that after spending entire last week on different things, just downgrading sharp and libvips solved the issue. We will put more logging to zero on the images which are causing issue and will give perf output.

Let me know if you want me to give any other data.

**Comment by kleisauke** [comment]
> * I am not sure if I have built libvips with loadable modules support. How can I check?

To check whether libvips was built with loadable modules support:
```bash
$ vips --vips-config
```
(look for `dynamic module: yes` and `enable modules: yes`)

Or in case the vips utility is not available:
```bash
$ grep "VIPS_CONFIG" $PREFIX/include/vips/version.h | cut -d'"' -f 2 | sed "s/, /\n/g"
```
(where `$PREFIX` is the directory where libvips was installed)

I guess the reason that you're building libvips yourself is to add additional images support (such as PDF or HEIC) and/or to compile with specific optimization flags. If so, can you tell me what additional libraries (+ version numbers) libvips was compiled with? And with what specific optimization flags (if any)?

**Comment by adityapatadia** [comment]
It shows `enable modules: yes` and in libheif section it shows `dynamic module: yes`. I use `-O3 -fPIC` as compiler flags. You are correct, I compile myself to have more libraries like HEIC. I custom compile libraries I mentioned above and rest of libraries like libgif, libde265, libpng, libtiff and pango and cairo are all just installed using `apt-get` of ubuntu focal.

We are getting more data on this and trying to zero on images which actually causes this.

**Comment by adityapatadia** [comment]
This is still a problem even after downgrade. We thought this was a problem with libvips but it seems something else. I will keep you posted as we investigate. You can close this or leave it open.

**Comment by jcupitt** [comment]
I'm sorry you're having problems, but I'm glad it's not libvips.

I'll close, please open a new issue if you find something else.

**Comment by adityapatadia** [comment]
The issue is that some of GIFs which are coming are very large in size and when given to sharp and eventually VIPS, they take very long time to process. I believe a timeout on processing will help a lot here. Currently there is no timeout.

**Comment by lovell** [comment]
@adityapatadia Please can you provide a sample image and minimal code that allows someone else to reproduce this. If you can reproduce it using vips at the command line, then please provide it here. If this only occurs with sharp, then please open a new issue on the sharp repo.

**Comment by jcupitt** [comment]
Yes, we're revising GIF handling at the moment, so a sample GIF that takes a long time to process would be helpful, if you have one.

I made a 30mb GIF for testing:

```
$ vipsheader x.gif[n=-1]
x.gif: 176x212800 uchar, 3 bands, srgb, gifload
```

It takes about 5s to thumbnail. If you have something even worse, I'd be very interested.

You can get the dimensions very quickly, so you can estimate processing time from that. You could reject GIFs with more than 100,000 pixels on an axis, for example.

As Lovell says, please open a new issue for slow GIF handling.

**Comment by adityapatadia** [comment]
![giphy(3)1546109899272](https://user-images.githubusercontent.com/1086617/133960736-0cc21414-5d8f-4809-add0-16a4d7c22246.gif)

Please check this GIF. It takes about 20k seconds 😁 I used mode=crop and cropped in aspect ratio of 2:1. Also tried to convert GIF to WebP.

**Comment by adityapatadia** [comment]
We also applied blur of 50.

**Comment by lovell** [comment]
A Gaussian blur with a large sigma of 50 is quite likely to be the bottleneck, but without code or a command line example that allows someone else to reproduce the problem, we're only able to guess. Perhaps you could use a tool like [callgrind](https://valgrind.org/docs/manual/cl-manual.html) to help you determine where CPU time is spent.

Running a full decode/encode round-trip at the command line via `vips copy` with the provided 120-frame 540x540 GIF image and the new cgif-based saver takes about 1.8 seconds.
